SABnzbd: use cross/7zz and cross/unrar7

hgy59 opened the pull request:
## Description

<!--Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context.-->
related to https://github.com/SynoCommunity/spksrc/pull/6356#issuecomment-2561708100

Add cross/7zz, cross/7za and cross/unrar7

### SABnzbd
- use cross/7zz 
- use cross/unrar7 for DSM 7
- remove cross/busybox dependency (use cross/coreutils for nice and cross/ionice instead)
- update par2cmdline-turbo

Fixes # <!--Optionally, add links to existing issues or other PR's-->

## Checklist

- [x] Build rule `all-supported` completed successfully
- [ ] New installation of package completed successfully
- [ ] Package upgrade completed successfully (Manually install the package again)
- [ ] Package [functionality was tested](https://github.com/SynoCommunity/spksrc/wiki/Package-Update-Policy#tests-checks)
- [ ] Any needed [documentation](https://github.com/SynoCommunity/spksrc/wiki/Create-documentation) is updated/created


### Type of change

<!--Please use any relevant tags.-->
- [x] New Libraries
- [x] Includes small framework changes

mreid-tt commented:
hey @hgy59, I note a build error because there is no dependent package. This I believe can be easily fixed with a conditional in the `build.yml` file similar to this https://github.com/dorny/paths-filter/issues/66. Would it be prudent to add the fix to this PR or should a separate one be submitted?

hgy59 commented:
> hey @hgy59, I note a build error because there is no dependent package. This I believe can be easily fixed with a conditional in the `build.yml` file similar to this [dorny/paths-filter#66](https://github.com/dorny/paths-filter/issues/66). Would it be prudent to add the fix to this PR or should a separate one be submitted?

@mreid-tt you can create a dedicated PR (I will add 7zip and unrar7 to spk/sabnzbd, so this will not be an issue here).

mreid-tt commented:
> @mreid-tt you can create a dedicated PR (I will add 7zip and unrar7 to spk/sabnzbd, so this will not be an issue here).

hey @hgy59, I've authored a dedicated PR as suggested and requested your review before merging. Tests with normal packages are successful with GitHub.

mreid-tt commented:
I've done a test install on DSM 7.2.2-72806 Update 2 on x64 and it installs okay. I've also done some test downloads to test out the new decompression binaries. Everything looks good so far...

Startup detects the new tools:
```
2025-01-03 09:40:12,838::INFO::[SABnzbd:460] UNRAR binary... found (/volume1/@appstore/sabnzbd/bin/unrar)
2025-01-03 09:40:12,839::INFO::[SABnzbd:468] UNRAR binary version 7.10
2025-01-03 09:40:12,840::INFO::[SABnzbd:475] 7za binary... found (/volume1/@appstore/sabnzbd/bin/7za)
2025-01-03 09:40:12,840::INFO::[SABnzbd:477] 7za binary version 24.09
```

Decompression using both of these work well:
```
2025-01-03 09:40:46,585::INFO::[postproc:475] Running unpacker on test_download_100MB
2025-01-03 09:40:46,586::INFO::[newsunpack:310] Unrar starting on /volume1/downloads/incomplete/test_download_100MB
2025-01-03 09:40:46,587::INFO::[newsunpack:564] Extracting rarfile /volume1/downloads/incomplete/test_download_100MB/sometestfile-100MB.part01.rar (belonging to sometestfile-100MB) to /volume1/downloads/complete/_UNPACK_test_download_100MB
2025-01-03 09:40:46,588::INFO::[misc:1262] [N/A] Running external command: ['/volume1/@appstore/sabnzbd/bin/nice', '-n15', '/volume1/@appstore/sabnzbd/bin/ionice', '-c2', '-n4', '/volume1/@appstore/sabnzbd/bin/unrar', 'x', '-idp', '-scf', '-o-', '-or', '-ai', '-p-', '/volume1/downloads/incomplete/test_download_100MB/sometestfile-100MB.part01.rar', '/volume1/downloads/complete/_UNPACK_test_download_100MB/']
2025-01-03 09:40:47,199::INFO::[newsunpack:873] Unpacked 2 files/folders in 0 seconds
2025-01-03 09:40:47,257::INFO::[newsunpack:314] Unrar finished on /volume1/downloads/incomplete/test_download_100MB
2025-01-03 09:40:47,259::INFO::[postproc:477] Unpacked files ['/volume1/downloads/complete/_UNPACK_test_download_100MB/100MB-explanation.txt', '/volume1/downloads/complete/_UNPACK_test_download_100MB/100MB.bin']
2025-01-03 09:40:47,264::INFO::[filesystem:341] Checking if any resulting filenames need to be sanitized
2025-01-03 09:40:47,267::INFO::[postproc:481] Finished unpack_magic on test_download_100MB
```
```
2025-01-03 09:47:37,459::INFO::[postproc:475] Running unpacker on Ultimate.Spiderman.S01E11.Venom.Attack.720p.WEB-DL.DD5.1.H264-NTb.mkv.7z
2025-01-03 09:47:37,469::INFO::[newsunpack:319] 7za starting on /volume1/downloads/incomplete/Ultimate.Spiderman.S01E11.Venom.Attack.720p.WEB-DL.DD5.1.H264-NTb.mkv.7z
2025-01-03 09:47:37,471::INFO::[newsunpack:901] Starting extract on 7zip set/file: Ultimate.Spiderman.S01E11.Venom.Attack.720p.WEB-DL.DD5.1.H264-NTb.mkv.7z 
2025-01-03 09:47:37,472::INFO::[misc:1262] [N/A] Running external command: ['/volume1/@appstore/sabnzbd/bin/nice', '-n15', '/volume1/@appstore/sabnzbd/bin/ionice', '-c2', '-n4', '/volume1/@appstore/sabnzbd/bin/7za', 'x', '-y', '-aou', '-ssc', '-p', '-o/volume1/downloads/complete/_UNPACK_Ultimate.Spiderman.S01E11.Venom.Attack.720p.WEB-DL.DD5.1.H264-NTb.mkv.7z', '/volume1/downloads/incomplete/Ultimate.Spiderman.S01E11.Venom.Attack.720p.WEB-DL.DD5.1.H264-NTb.mkv.7z/Ultimate.Spiderman.S01E11.Venom.Attack.720p.WEB-DL.DD5.1.H264-NTb.mkv.7z.001']
2025-01-03 09:48:49,255::INFO::[newsunpack:1011] Unpacked 1 files/folders in 1 min 11 seconds
2025-01-03 09:48:50,004::INFO::[newsunpack:323] 7za finished on /volume1/downloads/incomplete/Ultimate.Spiderman.S01E11.Venom.Attack.720p.WEB-DL.DD5.1.H264-NTb.mkv.7z
2025-01-03 09:48:50,021::INFO::[postproc:477] Unpacked files ['/volume1/downloads/complete/_UNPACK_Ultimate.Spiderman.S01E11.Venom.Attack.720p.WEB-DL.DD5.1.H264-NTb.mkv.7z/Ultimate.Spiderman.S01E11.Venom.Attack.720p.WEB-DL.DD5.1.H264-NTb.mkv']
2025-01-03 09:48:50,022::INFO::[filesystem:341] Checking if any resulting filenames need to be sanitized
2025-01-03 09:48:50,023::INFO::[postproc:481] Finished unpack_magic on Ultimate.Spiderman.S01E11.Venom.Attack.720p.WEB-DL.DD5.1.H264-NTb.mkv.7z
```

hgy59 commented:
@Safihre this PR is now an update of SABnzb dependencies.
Everything is prepared for a new package release, but we can hold it back, if an update of SABnzbd is planned for the near future (or you find some issues).

Can you please have a look at the packages created by github build action?
Beside the `par2` command, that shrinks from 7.8MB to 820KB due code optimization, the `sabctools-8.2.5` wheel shrinks from 630KB to 85KB.
The difference I can see in `sabctools-8.2.5`: the former (larger) was created with `bdist_wheel (0.44.0)` and the smaller with `setuptools (75.4.0)` (Property `Generator` in sabctools-8.2.5.dist-info\WHEEL). This must be related to the integration of #6282. 
Hopefully it is still fully functional.

Safihre commented on cross/7zz/Makefile:
Why also include 7zz? 
One of them is enough?

hgy59 replied:
7zz has additional capabilities like unrar and is the recommended tool under linux,
but sabnzbd does not support it (yet?)

Safihre replied:
We do support it, it's the first one we look for:
https://github.com/sabnzbd/sabnzbd/blob/7de08693765fac39192e4cbcffdcc725244a943e/sabnzbd/newsunpack.py#L148-L154
Could use that one instead?

hgy59 replied:
Yes we can.
does SABnzbd use 7zz to unpack rar or does it still require unrar?

Safihre replied:
Unrar is more optimized for Rar and since most of usenet content is rar'ed it makese sense that it's the main tool.

Safihre commented on cross/7za/PLIST:
```diff
@@ -0,0 +1 @@
+bin:bin/7za
```
For some reason SABnzbd isn't seeing it and jumping to the old system 7zip.
```
2025-01-13 22:37:33,269::INFO::[SABnzbd:460] UNRAR binary... found (/volume1/@appstore/sabnzbd/bin/unrar)
2025-01-13 22:37:33,269::INFO::[SABnzbd:468] UNRAR binary version 7.10
2025-01-13 22:37:33,270::INFO::[SABnzbd:475] 7za binary... found (/bin/7z)
2025-01-13 22:37:33,270::INFO::[SABnzbd:477] 7za binary version 16.02
```

We just look at the `PATH` and then check for any of the listed executable names in each part of the `PATH`.
https://github.com/sabnzbd/sabnzbd/blob/7de08693765fac39192e4cbcffdcc725244a943e/sabnzbd/newsunpack.py#L148-L154

hgy59 replied:
@Safihre thanks for the feedback, guess I fixed it.

Safihre reviewed (commented):
Nice work. Just for some reason it isn't picking up the new 7zip, but unrar is working.. Strange!
Tested on my arch64-7.1 system.

I promised a reward, so you can contact me on Discord so I can send it to you for the work :)

hgy59 commented:
Formats supported by 7za:
```
$ 7za/work-x64-7.1/install/usr/local/7za/bin/7za i

7-Zip (a) 24.09 (x64) : Copyright (c) 1999-2024 Igor Pavlov : 2024-11-29
 64-bit locale=C.UTF-8 Threads:4 OPEN_MAX:1048576


Formats:
   C...F..........c.a.m+..  7z       7z            7 z BC AF ' 1C
    ...F..................  Cab      cab           M S C F 00 00 00 00
    ......................  Split    001
   CK.....................  bzip2    bz2 bzip2 tbz2 (.tar) tbz (.tar) B Z h
   CK.................m+..  gzip     gz gzip tgz (.tar) tpz (.tar) apk (.tar) 1F 8B 08
    K.....O...............  lzma     lzma
    K.....................  lzma86   lzma86
   C......O...LH......m+..  tar      tar ova       offset=257 u s t a r
   CK.....................  xz       xz txz (.tar) FD 7 z X Z 00
   C...FMG........c.a.m+..  zip      zip z01 zipx jar xpi odt ods docx xlsx epub ipa apk appx P K 03 04  ||  P K 05 06  ||  P K 06 06  ||  P K 07 08 P K  ||  P K 0 0 P K
    K.....................  zstd     zst tzst (.tar) ( B5 / FD
   CK.....O.....XC........  Hash     sha256 sha512 sha384 sha224 sha3-256 sha1 sha md5 blake2sp xxh64 crc32 crc64 asc cksum

Codecs:
   4ED   303011B BCJ2
    EDF  3030103 BCJ
    EDF  3030205 PPC
    EDF  3030401 IA64
    EDF  3030501 ARM
    EDF  3030701 ARMT
    EDF  3030805 SPARC
    EDF        A ARM64
    EDF        B RISCV
    EDF    20302 Swap2
    EDF    20304 Swap4
    ED     40202 BZip2
    ED         0 Copy
    ED     40109 Deflate64
    ED     40108 Deflate
    EDF        3 Delta
    ED        21 LZMA2
    ED     30101 LZMA
    ED     30401 PPMD
    EDF  6F10701 7zAES
    EDF  6F00181 AES256CBC

Hashers:
      4        1 CRC32
     20      201 SHA1
     32        A SHA256
      8      211 XXH64
      8        4 CRC64
```

Formats supported by 7zz:
```
$ 7zz/work-x64-7.1/install/usr/local/7zz/bin/7zz i

7-Zip (z) 24.09 (x64) : Copyright (c) 1999-2024 Igor Pavlov : 2024-11-29
 64-bit locale=C.UTF-8 Threads:4 OPEN_MAX:1048576


Formats:
   C...F..........c.a.m+..  7z       7z            7 z BC AF ' 1C
    ......................  APFS     apfs img      offset=32 N X S B 00
    ......................  APM      apm           E R
    ......................  Ar       ar a deb udeb lib ! < a r c h > 0A
    ......................  Arj      arj           ` EA
    K.....O.....X.........  Base64   b64
    ......O...............  COFF     obj
    ...F..................  Cab      cab           M S C F 00 00 00 00
    ......................  Chm      chm chi chq chw I T S F 03 00 00 00 ` 00 00 00
    ......................  Compound msi msp doc xls ppt D0 CF 11 E0 A1 B1 1A E1
    ....M.................  Cpio     cpio          0 7 0 7 0  ||  C7 q  ||  q C7
    ......................  CramFS   cramfs        offset=16 C o m p r e s s e d 20 R O M F S
    .....G..B.............  Dmg      dmg           k o l y 00 00 00 04 00 00 02 00
    .........E............  ELF      elf           E L F
    ......................  Ext      ext ext2 ext3 ext4 img offset=1080 S EF
    ......................  FAT      fat img       offset=510 U AA
    ......................  FLV      flv           F L V 01
    ......................  GPT      gpt mbr       offset=512 E F I 20 P A R T 00 00 01 00
    ....M.................  HFS      hfs hfsx      offset=1024 B D  ||  H + 00 04  ||  H X 00 05
    ...F..................  Hxs      hxs hxi hxr hxq hxw lit I T O L I T L S 01 00 00 00 ( 00 00 00
    ......O...............  IHex     ihex
    ......................  Iso      iso img       offset=32769 C D 0 0 1
    ......................  LP       lpimg img     offset=4096 g D l a 4 00 00 00
    ......................  Lzh      lzh lha       offset=2 - l h
    .......P..............  MBR      mbr
    ....M....E............  MachO    macho         CE FA ED FE  ||  CF FA ED FE  ||  FE ED FA CE  ||  FE ED FA CF
    ......................  MsLZ     mslz          S Z D D 88 F0 ' 3 A
    ....M.................  Mub      mub           CA FE BA BE 00 00 00  ||  B9 FA F1 0E
    ......................  NTFS     ntfs img      offset=3 N T F S 20 20 20 20 00
    ...F.G................  Nsis     nsis          offset=4 EF BE AD DE N u l l s o f t I n s t
    .........E............  PE       exe dll sys   M Z
    ......................  Ppmd     pmd           8F AF AC 84
    ......................  QCOW     qcow qcow2 qcow2c Q F I FB 00 00 00
    ...F..................  Rar      rar r00       R a r ! 1A 07 00
    ...F..................  Rar5     rar r00       R a r ! 1A 07 01 00
    ......................  Rpm      rpm           ED AB EE DB
    K.....................  SWF      swf           F W S
    ....M.................  SWFc     swf (~.swf)   C W S  ||  Z W S
    ......................  Sparse   simg img      : FF & ED 01 00
    ......................  Split    001
    ....M.................  SquashFS squashfs      h s q s  ||  s q s h  ||  s h s q  ||  q s h s
    .........E............  TE       te            V Z
    ...FM.................  UEFIc    scap          BD 86 f ; v 0D 0 @ B7 0E B5 Q 9E / C5 A0  ||  8B A6 < J # w FB H 80 = W 8C C1 FE C4 M  ||  B9 82 91 S B5 AB 91 C B6 9A E3 A9 C F7 / CC
    ...FM.................  UEFIf    uefif         offset=16 D9 T 93 z h 04 J D 81 CE 0B F6 17 D8 90 DF  ||  x E5 8C 8C = 8A 1C O 99 5 89 a 85 C3 - D3
    ....M.O...............  Udf      udf iso img   offset=32768 00 B E A 0 1 01 00  ||  01 C D 0 0 1
    ......................  VDI      vdi           offset=64 10 DA BE
    .....G................  VHD      vhd           c o n e c t i x 00 00
    ......................  VHDX     vhdx avhdx    v h d x f i l e
    ......................  VMDK     vmdk          K D M V
    ......................  Xar      xar pkg xip   x a r ! 00
    ......................  Z        z taz (.tar)  1F 9D
   CK.....................  bzip2    bz2 bzip2 tbz2 (.tar) tbz (.tar) B Z h
   CK.................m+..  gzip     gz gzip tgz (.tar) tpz (.tar) apk (.tar) 1F 8B 08
    K.....O...............  lzma     lzma
    K.....................  lzma86   lzma86
   C......O...LH......m+..  tar      tar ova       offset=257 u s t a r
   C.SN.......LH..c.a.m+..  wim      wim swm esd ppkg M S W I M 00 00 00
   CK.....................  xz       xz txz (.tar) FD 7 z X Z 00
   C...FMG........c.a.m+..  zip      zip z01 zipx jar xpi odt ods docx xlsx epub ipa apk appx P K 03 04  ||  P K 05 06  ||  P K 06 06  ||  P K 07 08 P K  ||  P K 0 0 P K
    K.....................  zstd     zst tzst (.tar) ( B5 / FD
   CK.....O.....XC........  Hash     sha256 sha512 sha384 sha224 sha3-256 sha1 sha md5 blake2sp xxh64 crc32 crc64 asc cksum

Codecs:
   4ED   303011B BCJ2
    EDF  3030103 BCJ
    EDF  3030205 PPC
    EDF  3030401 IA64
    EDF  3030501 ARM
    EDF  3030701 ARMT
    EDF  3030805 SPARC
    EDF        A ARM64
    EDF        B RISCV
    EDF    20302 Swap2
    EDF    20304 Swap4
    ED     40202 BZip2
    ED         0 Copy
    ED     40109 Deflate64
    ED     40108 Deflate
    EDF        3 Delta
    ED        21 LZMA2
    ED     30101 LZMA
    ED     30401 PPMD
     D     40301 Rar1
     D     40302 Rar2
     D     40303 Rar3
     D     40305 Rar5
    EDF  6F10701 7zAES
    EDF  6F00181 AES256CBC

Hashers:
      4        1 CRC32
     16      208 MD5
     20      201 SHA1
     32        A SHA256
    256      231 SHA3-256
     48      222 SHA384
     64      223 SHA512
      8      211 XXH64
      8        4 CRC64
     32      202 BLAKE2sp
```

hgy59 commented:
> I promised a reward, so you can contact me on Discord so I can send it to you for the work :)

Thanks, I did this for free. You can sponser the SynoCommunity project instead (hopefully we will get a new respository server some day).

Safihre commented:
> > I promised a reward, so you can contact me on Discord so I can send it to you for the work :)
> 
> Thanks, I did this for free. You can sponser the SynoCommunity project instead (hopefully we will get a new respository server some day).

Donated $250.
https://github.com/sponsors/SynoCommunity

mreid-tt commented:
I've performed another test install on DSM 7.2.2-72806 Update 2 on x64 and it installs okay. I've also done some test downloads to test out the new decompression binaries.

Decompression using both of these work well:
```
2025-01-15 06:49:25,803::INFO::[postproc:475] Running unpacker on test_download_1000MB
2025-01-15 06:49:25,806::INFO::[newsunpack:310] Unrar starting on /volume1/downloads/incomplete/test_download_1000MB
2025-01-15 06:49:25,818::INFO::[newsunpack:564] Extracting rarfile /volume1/downloads/incomplete/test_download_1000MB/sometestfile-1GB.part01.rar (belonging to sometestfile-1GB) to /volume1/downloads/complete/_UNPACK_test_download_1000MB
2025-01-15 06:49:25,820::INFO::[misc:1262] [N/A] Running external command: ['/volume1/@appstore/sabnzbd/bin/nice', '-n15', '/volume1/@appstore/sabnzbd/bin/ionice', '-c2', '-n4', '/volume1/@appstore/sabnzbd/bin/unrar', 'x', '-idp', '-scf', '-o-', '-or', '-ai', '-p-', '/volume1/downloads/incomplete/test_download_1000MB/sometestfile-1GB.part01.rar', '/volume1/downloads/complete/_UNPACK_test_download_1000MB/']
2025-01-15 06:49:41,127::INFO::[newsunpack:873] Unpacked 2 files/folders in 15 seconds
2025-01-15 06:49:41,258::INFO::[newsunpack:314] Unrar finished on /volume1/downloads/incomplete/test_download_1000MB
2025-01-15 06:49:41,288::INFO::[postproc:477] Unpacked files ['/volume1/downloads/complete/_UNPACK_test_download_1000MB/1GB-explanation.txt', '/volume1/downloads/complete/_UNPACK_test_download_1000MB/1GB.bin']
2025-01-15 06:49:41,291::INFO::[filesystem:341] Checking if any resulting filenames need to be sanitized
2025-01-15 06:49:41,294::INFO::[postproc:481] Finished unpack_magic on test_download_1000MB
```
```
2025-01-15 06:50:58,847::INFO::[postproc:475] Running unpacker on Ultimate.Spiderman.S01E08.Back.in.Black.720p.WEB-DL.DD5.1.H264-NTb.7z
2025-01-15 06:50:58,854::INFO::[newsunpack:319] 7za starting on /volume1/downloads/incomplete/Ultimate.Spiderman.S01E08.Back.in.Black.720p.WEB-DL.DD5.1.H264-NTb.7z
2025-01-15 06:50:58,856::INFO::[newsunpack:901] Starting extract on 7zip set/file: Ultimate.Spiderman.S01E08.Back.in.Black.720p.WEB-DL.DD5.1.H264-NTb.7z 
2025-01-15 06:50:58,857::INFO::[misc:1262] [N/A] Running external command: ['/volume1/@appstore/sabnzbd/bin/nice', '-n15', '/volume1/@appstore/sabnzbd/bin/ionice', '-c2', '-n4', '/volume1/@appstore/sabnzbd/bin/7zz', 'x', '-y', '-aou', '-ssc', '-p', '-o/volume1/downloads/complete/_UNPACK_Ultimate.Spiderman.S01E08.Back.in.Black.720p.WEB-DL.DD5.1.H264-NTb.7z', '/volume1/downloads/incomplete/Ultimate.Spiderman.S01E08.Back.in.Black.720p.WEB-DL.DD5.1.H264-NTb.7z/Ultimate.Spiderman.S01E08.Back.in.Black.720p.WEB-DL.DD5.1.H264-NTb.7z.001']
2025-01-15 06:52:15,471::INFO::[newsunpack:1011] Unpacked 1 files/folders in 1 min 16 seconds
2025-01-15 06:52:15,995::INFO::[newsunpack:323] 7za finished on /volume1/downloads/incomplete/Ultimate.Spiderman.S01E08.Back.in.Black.720p.WEB-DL.DD5.1.H264-NTb.7z
2025-01-15 06:52:16,085::INFO::[postproc:477] Unpacked files ['/volume1/downloads/complete/_UNPACK_Ultimate.Spiderman.S01E08.Back.in.Black.720p.WEB-DL.DD5.1.H264-NTb.7z/Ultimate.Spiderman.S01E08.Back.in.Black.720p.WEB-DL.DD5.1.H264-NTb.mkv']
2025-01-15 06:52:16,086::INFO::[filesystem:341] Checking if any resulting filenames need to be sanitized
2025-01-15 06:52:16,086::INFO::[postproc:481] Finished unpack_magic on Ultimate.Spiderman.S01E08.Back.in.Black.720p.WEB-DL.DD5.1.H264-NTb.7z
```
As for the startup, this reports the following detections:
```
2025-01-15 06:46:06,893::INFO::[SABnzbd:460] UNRAR binary... found (/volume1/@appstore/sabnzbd/bin/unrar)
2025-01-15 06:46:06,893::INFO::[SABnzbd:468] UNRAR binary version 7.10
2025-01-15 06:46:06,894::INFO::[SABnzbd:475] 7za binary... found (/volume1/@appstore/sabnzbd/bin/7zz)
2025-01-15 06:46:06,894::INFO::[SABnzbd:477] 7za binary version 24.09
```
Note that it reports `7za binary` even though it finds `7zz`. This may be a cosmetic issue in SABnzbd to fix upstream.

Safihre commented:
It will still be a while before the next Sab release, so I think this shouldn't wait for that.
